libdrgn/stack_trace.c

```diff
@@ -1106,9 +1114,31 @@ drgn_object_stack_trace(const struct drgn_object *obj,
 		if (err)
 			return err;
 		return drgn_get_stack_trace(drgn_object_program(obj),
-					    value.uvalue, NULL, ret);
+					    value.uvalue, NULL, NULL, ret);
 	} else {
-		return drgn_get_stack_trace(drgn_object_program(obj), 0, obj,
+		return drgn_get_stack_trace(drgn_object_program(obj), 0, obj, NULL,
 					    ret);
 	}
 }
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_thread_stack_trace(struct drgn_thread *thread,
+			struct drgn_stack_trace **ret)
+{
+	if (thread->prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+		struct nstring *prstatus = NULL;
+		struct drgn_thread *crashed_thread;
+		if (!(thread->prog->flags & DRGN_PROGRAM_IS_LIVE)) {
+			struct drgn_error *err;
+			err = drgn_program_crashed_thread(thread->prog,
+							  &crashed_thread);
+			if (!err && crashed_thread->tid == thread->tid)
```

Comment:
This isn't correct if these are two different idle threads; they'll both have TID 0, but they're different threads. Also, this is leaking `err`. I'd just get rid of this `crashed_thread` lookup entirely and just use `thread->prstatus` directly. It'll be set if `thread` is `crashed_thread`, and otherwise we can just look it up again.

Reply:
`thread->prstatus` is a value, not a pointer, so it'll actually hold garbage if `thread` isn't `crashed_thread`, hence why I was checking. Given that it just takes a shallow copy everywhere though, I suppose there's no reason why I can't just make it a pointer instead and then do what you've suggested.

Reply:
I think another option would be to initialize `prstatus` to an empty string (`.str = NULL, .len = 0`), would that work?